src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs

```diff
@@ -84,68 +82,92 @@ public async Task<bool> PushMetadataAsync(CancellationToken cancellationToken)
-                    // upload it to the BlobArtifacts folder only when publishingVersion >= 3
-                    if (manifestBuildData.PublishingVersion >= 3)
+                    List<SigningInformation> signingInformation = new List<SigningInformation>();
+                    foreach (var m in parsedManifest)
                     {
-                        SigningInformation finalSigningInfo = MergeSigningInfo(signingInformation);
+                        if (m.SigningInformation != null)
+                        {
+                            signingInformation.Add(m.SigningInformation);
+                        }
+                    }
 
-                        // Inject an entry of MergedManifest.xml to the in-memory merged manifest
-                        string location = null;
-                        AssetData assetData = finalBuild.Assets.FirstOrDefault();
+                    //get packages blobs and signing info 
+                    (List<PackageArtifactModel> packages,
+                        List<BlobArtifactModel> blobs) = GetPackagesAndBlobsInfo(manifest);
 
-                        if (assetData != null)
-                        {
-                            AssetLocationData assetLocationData = assetData.Locations.FirstOrDefault();
+                    //create merged buildModel to create the merged manifest
+                    BuildModel modelForManifest =
+                        CreateMergedManifestBuildModel(packages, blobs, manifest);
 
-                            if (assetLocationData != null)
-                            {
-                                location = assetLocationData.Location;
-                            }
-                        }
 
-                        finalBuild.Assets = finalBuild.Assets.Add(GetManifestAsAsset(finalBuild.Assets, location, MergedManifestFileName));
+                    //add manifest as an asset to the buildModel
+                    var mergedManifestAsset = GetManifestAsAsset(blobs, MergedManifestFileName);
+                    modelForManifest.Artifacts.Blobs.Add(mergedManifestAsset);
+                    
+                    SigningInformation finalSigningInfo = MergeSigningInfo(signingInformation);
 
-                        BuildModel modelForManifest = CreateMergedManifestBuildModel(finalBuild.Assets, manifestBuildData);
+                    // push the merged manifest, this is required for only publishingVersion 3 and above
+                    if (manifest.PublishingVersion >= 3)
+                    {
                         PushMergedManifest(modelForManifest, finalSigningInfo);
                     }
 
-                    Client.Models.Build recordedBuild = await client.Builds.CreateAsync(finalBuild, cancellationToken);
+                    // populate buildData and assetData using merged manifest data 
+                    BuildData buildData = UpdateBuildDataFromMergedManifest(modelForManifest, manifest, cancellationToken);
```

Comment:
Nit: Since `buildData` doesn't exist prior to this call, perhaps it should be `GetMaestroBuildDataFromMergedManifest`?

Reply:
done